Fix ZStandard native memory leaks

lorban opened the pull request:
Fixes #14028 by closing the zstd context when sources / sinks are released or when the GC cleans them up.

Also add buffer pooling to the zstd streams.

lorban committed:
#14028 close context in release() to release all native resources

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

joakime reviewed (dismissed):
Simple enough.

lorban committed:
enable buffer pooling on zstd streams

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
fix NPE when written content is last and null

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
add phantom ref cleaner to make sure zstd contexts do release their native resources even when the sink does not do a last write or the source is not read until EOF nor failed

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
fix checkstyle

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
change default bufferSizeCeiling to match the default ArrayByteBufferPool maxCapacity

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

joakime commented on jetty-core/jetty-compression/jetty-compression-zstandard/src/main/java/org/eclipse/jetty/compression/zstandard/ZstandardDecoderConfig.java:
```diff
@@ -35,7 +35,7 @@ public class ZstandardDecoderConfig implements DecoderConfig
         // but put some upper limit on it for our default buffer size.
         // The user can still configure the buffer size to be higher if they want to.
-        long bufferSizeCeiling = 256_000;
+        long bufferSizeCeiling = 64 * 1024L; // TODO this is the default ArrayByteBufferPool maxCapacity, there should be a way to discover that
```
This size is smaller than the recommended size by the zstandard lib.

joakime commented on jetty-core/jetty-compression/jetty-compression-zstandard/src/main/java/org/eclipse/jetty/compression/zstandard/ZstandardEncoderConfig.java:
```diff
@@ -36,7 +36,7 @@ public class ZstandardEncoderConfig implements EncoderConfig
         // but put some upper limit on it for our default buffer size.
         // The user can still configure the buffer size to be higher if they want to.
-        long bufferSizeCeiling = 256_000;
+        long bufferSizeCeiling = 64 * 1024L; // TODO this is the default ArrayByteBufferPool maxCapacity, there should be a way to discover that
```
This size is smaller than the recommended size by the zstandard lib.

joakime reviewed (changes requested):
I don't like the reduction in recommended default size.

Can we instead have a log warning be produced if the bufferPool isn't setup properly to allow for zstandard lib recommendations?

lorban commented:
@joakime we currently have no API to detect the maximum pooled buffer size, so if we were to warn "please make  sure your pool is properly configured" we'd always do it, even when the pool is properly configured.

I hardcoded 64 KB here because the default maximum pooled buffer size is 64 KB, and it is the only solution that is both quick and easy to do that I could think of.

If zstd were to use non-pooled buffer, the performance would go down the drain: each direct byte buffer object has to be finalized (requires enqueuing/dequeuing from a concurrent queue, expensive!) and each time a new buffer is allocated it has to be zero'ed (filling 132 KB of memory with zeros, expensive!).

I agree this solution isn't ideal, but the only alternative I can come up that doesn't require modifying the `ByteBufferPool` interface with is to change the default `ArrayByteBufferPool` implementation to pool up to 256 KB buffers, i.e.: the previous ceiling value.

@sbordet @gregw opinions?

lorban committed:
restore bufferSizeCeiling to 256 KB + modify ABBP to pool up to 256 KB buffers

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
remove unneeded Compression.getBufferSize(), make zstd use the buffer size from its config instead like gzip and brotli already do

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
restore bufferSizeCeiling to 256 KB + modify ABBP to pool up to 256 KB buffers

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java:
```diff
@@ -66,7 +66,7 @@
 {
     static final int DEFAULT_FACTOR = 4096;
-    static final int DEFAULT_MAX_CAPACITY_BY_FACTOR = 16;
+    static final int DEFAULT_MAX_CAPACITY_BY_FACTOR = 64;
```
I don't like this change to the generic component (the buffer pool) driven by a component (zstd) that may not even be used.

I think it should be the other way around.

I would rather document the zstd `*Config.setBufferSize()` methods so that they refer back to the `ByteBufferPool`, saying that whatever size is set, it should fall within the limits of the `ByteBufferPool`.

sbordet commented:
Reading https://datatracker.ietf.org/doc/html/rfc8878#section-3.1.1.1.2, the min window size can be just 1 KiB.
However, there is a recommendation for 8 MiB, which seems just huge.

I think it's best for us to stick with a smaller size like 64 KiB, and document the implications with the `ByteBufferPool`.

lorban committed:
revert to capping the zstd buffer size to 64 KB by default, and javadoc ByteBufferPool implications of changing the default

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban commented:
So I've reverted to capping the zstd buffer sizes to 64 KB, and I added a comment to the javadoc explaining that this setting has implications on the `ByteBufferPool` configuration.

I believe this is the most reasonable thing to do for now as a user can always manually set the buffer size.

lorban committed:
revert test too

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

joakime commented:
The zstandard requirements are `Window_Size + 2 * Block_Maximum_Size`

Of which, ...

* `Window_Size` defaults to 128 KB.
* `Block_Maximum_Size = min(128 KB, Window_Size)`.

https://github.com/facebook/zstd/wiki/Using-libzstd-in-a-memory-constrained-environment

joakime commented:
We use the `zstd-jni` which is just a Java JNI layer on top of the facebook zstandard lib.
That lib has updated real-world defaults of 128 KB for Window Size.

It recommends using 128KB for the Window Size if dealing with sub 2GB inputs on non-streaming modes (we don't use the streaming modes).
It recommends higher Windows Size of up to 8MB if dealing with inputs over 2GB.

lachlan-roberts commented on jetty-core/jetty-compression/jetty-compression-common/src/main/java/org/eclipse/jetty/compression/Compression.java:
```diff
@@ -82,18 +81,6 @@ public String etag(String etag)
         return etag + etagSuffix;
     }
 
-    public int getBufferSize()
```
This is public API so we should not remove these methods in a dot release, these should be deprecated instead.

lorban replied:
Good catch, let me restore them and document them as deprecated.

lorban committed:
restore API methods + document them as deprecated.

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
resort to documenting the buffer size issue for now

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
remove useless call

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban commented:
@joakime @sbordet I've reverted the code changes related to the zstd buffer size and left the issue as-is for now, with only changes to the javadoc to hint at the problem.

This isn't a regression as this problem exists since the introduction of zstd (de)compression and we need the other fixes to be included in the next release that has to be cut this week, so let's postpone the buffer problem to the next release and simply ignore it for now.

lorban committed:
do not use zstd buffer logic for the default buffer size as zstd uses its own buffer internally and we only use the configured buffer size to control the buffer receiving (de)compressed data + fix bug in continueOp() buffer rollback logic

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

joakime commented on jetty-core/jetty-compression/jetty-compression-zstandard/src/main/java/org/eclipse/jetty/compression/zstandard/ZstandardDecoderConfig.java:
```diff
@@ -15,38 +15,12 @@
-    }
-
-    private int bufferSize = DEFAULT_BUFFER_SIZE;
+    private int bufferSize = org.eclipse.jetty.util.IO.DEFAULT_BUFFER_SIZE;
```
This is an invalid configuration on it's own.
The IO.DEFAULT_BUFFER_SIZE value is currently at 8192.
That's below the 128KB Window Size that you didn't change.

This WILL cause problems in zstandard lib.

Note that 256_000 is the ceiling that we implement for buffer size.
We stick with the recommended size from the zstandard lib implementation (from a call to `ZstdInputStreamNoFinalizer.recommendedDOutSize()`, which is currently at the value of `131591` with this version of the lib)

The logic in the original code needs to exist, somewhere, just to ensure that valid values are being used.
Probably on the .setBufferSize() call, to warn if setting a buffer size lower than recommended.

If we want a smaller default buffer size, then we also need to tune the Window Size Log, with calls to both ...

* `ZstdOutputStreamNoFinalizer.setLong(int)`
* `ZstdOutputStreamNoFinalizer.setWindowLog(int)`

Note, the two settings above are log values, not direct byte sizes.

This is not as simple as setting a small buffer size, zstandard needs many knobs tuned to achieve this end goal.

lorban replied:
How is that `bufferSize` setting impacting the zstd buffer? AFAICT it's only used as the output to copy decompressed bytes, the zstd lib still uses a buffer with a size of its choice internally.

joakime replied:
Sorry, `ZstdOutputStreamNoFinalizer` is for Encoding, not Decoding.

joakime reviewed (changes requested):
The current PR is invalid for zstandard, and will break it.

joakime commented on jetty-core/jetty-compression/jetty-compression-gzip/src/test/java/org/eclipse/jetty/compression/gzip/AbstractGzipTest.java:
```diff
@@ -95,15 +95,8 @@ public void tearDown()
     }
 
     protected void startGzip() throws Exception
```
Don't change the behavior outside of zstandard in this PR.

joakime commented on jetty-core/jetty-compression/jetty-compression-gzip/src/test/java/org/eclipse/jetty/compression/gzip/GzipCompressionTest.java:
```diff
@@ -35,7 +35,6 @@ public class GzipCompressionTest extends AbstractGzipTest
     public void testBigBlockWithExtraBytesViaGzipInputStream() throws Exception
     {
-        startGzip(64);
         String data1 = "0123456789ABCDEF".repeat(10);
```
You broke the purpose of these test cases.
These are historical test cases for small buffer size configurations that have caused us problems in the past.
Revert these changes to Gzip, and the base EncoderSink

lorban replied:
`startGzip` was initializing a `GzipCompression` instance that was not used in this test. So if that was meant to control a buffer size, the test was bogus to being with.

lorban committed:
fix position rollback

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
handle review comments

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
handle review comments

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

joakime reviewed (dismissed):
Good, for now.
We need a new issue to track the ByteBufferPool changes to make it expose it's limits.

lorban committed:
Merge remote-tracking branch 'origin/jetty-12.1.x' into fix/12.1.x/14028-native-leak-zstd

lorban committed:
fix checkstyle

Signed-off-by: Ludovic Orban <lorban@bitronix.be>
